File: snippets/cpp/dsu.cpp
struct DSU {
    vi p, sz;
    DSU(int n) : p(n), sz(n, 1) { iota(all(p), 0); }
    int find(int x) { return p[x] == x ? x : p[x] = find(p[x]); }
    bool unite(int x, int y) {
        x = find(x), y = find(y);
        if (x == y) return false;
        if (sz[x] < sz[y]) swap(x, y);
        p[y] = x, sz[x] += sz[y];
        return true;
    }
    bool same(int x, int y) { return find(x) == find(y); }
};

File: snippets/rust/dsu.rs
struct DSU {
    p: Vec<usize>,
    sz: Vec<usize>,
}
impl DSU {
    fn new(n: usize) -> Self { Self { p: (0..n).collect(), sz: vec![1; n] } }
    fn find(&mut self, x: usize) -> usize {
        if self.p[x] == x { x } else {
            let r = self.find(self.p[x]);
            self.p[x] = r;
            r
        }
    }
    fn unite(&mut self, x: usize, y: usize) -> bool {
        let mut x = self.find(x);
        let mut y = self.find(y);
        if x == y { return false; }
        if self.sz[x] < self.sz[y] { std::mem::swap(&mut x, &mut y); }
        self.p[y] = x;
        self.sz[x] += self.sz[y];
        true
    }
    fn same(&mut self, x: usize, y: usize) -> bool { self.find(x) == self.find(y) }
    fn size_of(&mut self, x: usize) -> usize { let r = self.find(x); self.sz[r] }
}
